### ISSUE TITLE
Covariance matrices not symmetric

### ISSUE THREAD

**Issue by MacDaddio** [borchero/pycave#23]
I observed that some of the covariance matrices are not symmetric. Most of them are close but some are considerably off. It turns out that multiplying the precisions_cholesky matrices and then taking an inverse was causing the issue. If you take the inverse and then multiply the resulting inverse precisions_cholesky matrices, then the covariances come out as symmetric. It does take a little longer (about 20%) for some reason but I think this would be a more accurate way to calculate the covariances. Let me know if this makes sense or if I am missing anything important.

Minimal working example:

from pycave.bayes import GaussianMixture
import torch
import numpy as np
import pycave
import random
import time

#Set seed
seed = 0
torch.manual_seed(seed)
torch.cuda.manual_seed(seed)
random.seed(seed)
    
#Inputs
n = 10000
p = 2000
k = 40

#Make some Gaussian data
X = torch.randn(n,p)

#Fit PyCave GMM
gmm = GaussianMixture(num_components=k, 
                      covariance_type='full', 
                      init_strategy='kmeans',
                      trainer_params={'gpus':1,'enable_progress_bar':False},
                      covariance_regularization=1e-3)
gmm = gmm.fit(X)

# NOT SYMMETRIC - Show the top corner of the covariance matrix:
t_start = time.time()
covs = gmm.model_.covariances
t_end = time.time()
print('Time to compute covariance =', t_end - t_start)
print(covs[0,0:5,0:5])

# SYMMETRIC - Show the top corner of the covariance matrix when calculated differently
t_start = time.time()
P = torch.linalg.inv(gmm.model_.precisions_cholesky)
covs2 = torch.bmm(P.transpose(1, 2), P)
t_end = time.time()
print('Time to compute covariance =', t_end - t_start)
print(covs2[0,0:5,0:5])

**Comment by borchero** [comment]
Nice catch!! This also allows me to decrease the tolerance in the tests. Thank you!
